src/platform/datapath_raw_xdp_win.c

```diff
@@ -107,246 +108,137 @@ CxPlatXdpExecute(
-                    free(rssTable);
-                    hRes = QUIC_STATUS_OUT_OF_MEMORY;
-                    goto Cleanup;
-                }
+        XSK_RING_INFO_SET TxRingInfo;
+        uint32_t TxRingInfoSize = sizeof(TxRingInfo);
+        Status = Xdp->XdpApi->XskGetSockopt(TxXsk, XSK_SOCKOPT_RING_INFO, &TxRingInfo, &TxRingInfoSize);
+        if (QUIC_FAILED(Status)) { CloseHandle(TxXsk); return Status; }
 
-                hr = obj->lpVtbl->Get(obj, L"ProcessorNumber", 0, &vtProp, 0, 0);
-                UINT32 procNum = vtProp.iVal;
-                VariantClear(&vtProp);
-                hr = obj->lpVtbl->Get(obj, L"ProcessorGroup", 0, &vtProp, 0, 0);
-                UINT32 groupNum = vtProp.iVal;
-                VariantClear(&vtProp);
-                CXPLAT_DBG_ASSERT(groupNum < numberOfProcGroups);
-                CXPLAT_DBG_ASSERT(procNum < numberOfProcs);
-                *(rssTable + groupNum * numberOfProcs + procNum) = 1;
-                obj->lpVtbl->Release(obj);
-            }
+        XSK_RING TxRing, TxCompletionRing;
+        XskRingInitialize(&TxRing, &TxRingInfo.Tx);
+        XskRingInitialize(&TxCompletionRing, &TxRingInfo.Completion);
 
-            SafeArrayUnaccessData(pSafeArray);
+        uint32_t TxIndex;
+        XskRingProducerReserve(&TxRing, MAXUINT32, &TxIndex);
 
-            // Count unique RSS procs by counting ones in rssTable.
-            for (DWORD i = 0; i < rssTableSize; ++i) {
-                cnt += rssTable[i];
-            }
+        XSK_BUFFER_DESCRIPTOR* Buffer = XskRingGetElement(&TxRing, TxIndex++);
+        Buffer->Address.BaseAddress = 0;
+        Buffer->Address.Offset = FIELD_OFFSET(XDP_TX_PACKET, FrameBuffer);
+        Buffer->Length = TxPacket.Buffer.Length;
+        XskRingProducerSubmit(&TxRing, 1);
 
-            free(rssTable);
-        }
+        XSK_NOTIFY_RESULT_FLAGS OutFlags;
+        Status = Xdp->XdpApi->XskNotifySocket(TxXsk, XSK_NOTIFY_FLAG_POKE_TX|XSK_NOTIFY_FLAG_WAIT_TX, 1000, &OutFlags);
+        if (QUIC_FAILED(Status)) { CloseHandle(TxXsk); return Status; }
 
-        VariantClear(&vtProp);
-        pclsObj->lpVtbl->Release(pclsObj);
-    }
+        uint32_t CompIndex;
+        if (XskRingConsumerReserve(&TxCompletionRing, MAXUINT32, &CompIndex) == 0) {
+            CloseHandle(TxXsk);
+            return E_ABORT;
+        }
+        XskRingConsumerRelease(&TxCompletionRing, 1);
 
-    //AF_XDP_LOG(INFO, "counted %u active queues on %s\n", cnt, if_name);
-    *Count = cnt;
+        PROCESSOR_NUMBER ProcNumber;
+        uint32_t ProcNumberSize = sizeof(PROCESSOR_NUMBER);
+        Status = Xdp->XdpApi->XskGetSockopt(TxXsk, XSK_SOCKOPT_TX_PROCESSOR_AFFINITY, &ProcNumber, &ProcNumberSize);
+        if (QUIC_FAILED(Status)) { CloseHandle(TxXsk); return Status; }
 
-Cleanup:
+        const CXPLAT_PROCESSOR_GROUP_INFO* Group = &CxPlatProcessorGroupInfo[ProcNumber.Group];
+        Queues[i] = Group->Offset + (ProcNumber.Number % Group->Count);
```

Comment:
why is the modulo necessary? i probably just need to go read more of the QUIC code, but i would have expected all ProcNumber.Numbers to be less than Group->Count.

Reply:
I _think_ it shouldn't be necessary.

Reply:
is it for hot add? where QUIC has a fixed set of CPUs per group that might see more CPUs added to the group later?

Reply:
No, we fix this list at process start.

Reply:
right, so if the group is fixed at size 1 during process start, and the OS later adds proc 2 to the group, then that's why you have the modulo?